ts/project.ts

```diff
@@ -143,17 +162,36 @@ const proxyEnvironmentHelpers = () => ({
     return this.defaultEnvironment.build(s, ...args);
   },
 
+  addExecutable<T extends Project, Name extends string>(this: T, name: Name) {
+    if (this.defaultEnvironment == null) {
+      throw new Error(
+        `'.addExecutable' can only be called on projects with a default environment`
+      );
+    }
+    return (s: TemplateStringsArray, ...args: Array<string>) => {
+      const newExecutable = { [name]: this.shell(s, ...args) } as {
+        [n in Name]: Executable;
+      };
+      return {
+        ...this,
+        ...newExecutable,
+      };
+    };
+  },
+
   addCheck<T extends Project, Name extends string>(this: T, name: Name) {
     if (this.defaultEnvironment == null) {
       throw new Error(
         `'.addCheck' can only be called on projects with a default environment`
       );
     }
     return (s: TemplateStringsArray, ...args: Array<string>) => {
-      const newCheck = this.check(s, ...args);
+      const newCheck = { [name]: this.check(s, ...args) } as {
+        [n in Name]: Check;
+      };
```

Comment:
The `as` cast is a bit sad, but luckily typescript will error if the type isn't close (e.g. we expect a `{ [Name]: Project }` instead of `{ [Name]: Check }` here would require an `as any as ...`.

I'm not quite sure why typescript was losing the type here, but while `name` is correctly typed as `Name`, `{ [name]: Check }` was becoming into `{ [string]: Check }`. I might keep looking into this this weekend to figure out why this is happening

Reply:
Thanks for explaining. Yeah, typescript sometimes assumes type `string` and often that's whet we want, but sometimes not...